criu/sk-udp.c

```diff
@@ -0,0 +1,655 @@
+#include <netinet/udp.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#include <sys/uio.h>
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <linux/sockios.h>
+
+#include "../soccr/soccr.h"
+#include "xmalloc.h"
+#include "sk-inet.h"
+#include "netfilter.h"
+#include "namespaces.h"
+
+#include "protobuf.h"
+#include "images/udp-queue.pb-c.h"
+
+struct udp_packet {
+	UdpPacketEntry *upe;
+	char *data;
+	struct list_head packet_list;
+};
+
+static LIST_HEAD(recv_queue_packets);
+/* The send queue only contains 1 packet */
+static struct udp_packet *send_queue_packet = NULL;
```

Comment:
I've already wrote about that. Using global variables like that is not safe.

Let's take a look on `send_queue_packet` which is used in two functions
`dump_send_queue` and `dump_one_udp`.
`dump_one_udp` calls `dump_send_queue`.

Stack:
`dump_one_udp`
`do_dump_one_inet_fd`
`dump_one_inet_fd` (->dump callback on `struct fdtype_ops`)
`do_dump_gen_file`
`dump_socket`
`dump_one_file`
`dump_task_files_seized` / `dump_my_file` (used in unix sock scm's support when transferring fd through unixsk)
for_each_pstree_item `dump_one_task`
`cr_dump_tasks`

As for now we dump file descriptors sequentially, so, no bug here. But I think it's better to move `send_queue_packet` and `g_data` to local variables of `dump_one_udp` to be on the safe side. The same about list `recv_queue_packets`.

Now let's analyze `udp_repair_sockets`. Restore stage of files is done in parallel way, so we need to have locking here. If you decide that udp sk network locking is needed and keep `udp_repair_sockets` list, then looks like we need to protect this list by mutex.

Reply:
I changed these variables into local variables of `dump_one_udp` in the latest update